core/primitives-core/src/account.rs

```diff
@@ -86,299 +116,236 @@ impl Account {
-struct AccountV2 {
+/// Account representation for serde ser/deser that maintains both backward
+/// and forward compatibility.
+#[derive(serde::Serialize, serde::Deserialize, PartialEq, Eq, Debug, Clone, ProtocolSchema)]
+struct SerdeAccount {
+    #[serde(with = "dec_format")]
     amount: Balance,
+    #[serde(with = "dec_format")]
     locked: Balance,
+    #[serde(with = "dec_format", default, skip_serializing_if = "Option::is_none")]
+    permanent_storage_bytes: Option<StorageUsage>,
     code_hash: CryptoHash,
     storage_usage: StorageUsage,
-    #[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
-    permanent_storage_bytes: StorageUsage,
+    /// Version of Account in re migrations and similar.
+    #[serde(default)]
+    version: AccountVersion,
 }
 
-/// We need custom serde deserialization in order to parse mainnet genesis accounts (LegacyAccounts)
-/// as accounts V1. This preserves the mainnet genesis hash.
-#[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
 impl<'de> serde::Deserialize<'de> for Account {
     fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
         D: serde::Deserializer<'de>,
     {
-        #[derive(serde::Deserialize)]
-        struct AccountData {
-            #[serde(with = "dec_format")]
-            amount: Balance,
-            #[serde(with = "dec_format")]
-            locked: Balance,
-            // If the field is missing, serde will use None as the default.
-            #[serde(default, with = "dec_format")]
-            permanent_storage_bytes: Option<StorageUsage>,
-            code_hash: CryptoHash,
-            storage_usage: StorageUsage,
-            #[serde(default)]
-            version: Option<AccountVersion>,
-        }
-
-        let account_data = AccountData::deserialize(deserializer)?;
-
-        match account_data.permanent_storage_bytes {
-            Some(permanent_storage_bytes) => {
-                // Given that the `permanent_storage_bytes` field has been serialized, the `version` field must has been serialized too.
-                let version = match account_data.version {
-                    Some(version) => version,
-                    None => {
-                        return Err(serde::de::Error::custom("missing `version` field"));
-                    }
-                };
-
-                #[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
-                if version < AccountVersion::V2 && permanent_storage_bytes > 0 {
+        let account_data = SerdeAccount::deserialize(deserializer)?;
+        match account_data.version {
+            AccountVersion::V1 => {
+                if account_data.permanent_storage_bytes.is_some() {
                     return Err(serde::de::Error::custom(
-                        "permanent storage bytes positive amount exists for account version older than V2",
+                        "permanent storage bytes exists for V1 account",
```

Comment:
nit: the err message is a bit unclear
```suggestion
                        "permanent storage bytes should not be set for V1 account",
```

Reply:
good point, thanks!